client/iotcore/client.go

```diff
@@ -330,7 +330,7 @@ func (c *client) DeleteDevice(
 			}
 		}
 		// Thing Principle still not detached; return error
-		if len(respListThingPrincipals.Principals) > 0 {
+		if respListThingPrincipals != nil && len(respListThingPrincipals.Principals) > 0 {
```

Comment:
The added condition should be checked for earlier (line 285) as there are other branches that also uses `respListThingPrincipals.Principals` without checking.

Reply:
I'm not sure you are right; at line 321 it overwrites `respListThingPrincipals` and this is where it can become `nil`

Reply:
Ohh, you're right!
My mistake. I believe `ListThingPrincipals` should never return `nil, nil`. However, the error from line 321 is never used apart from breaking the loop (at least logging it could be useful for debugging purposes).